### PR TITLE
Feature fields encoding

### PR DESCRIPTION
**Pull request by drons**
Hello!
When reading string data from tables it is not known in which encoding they are returned.
But the encoding is important in the further processing of data.
For now, i am add 'ENCODING' metadata to the layers for the MITAB & ESRI drivers Shapefile formats.

### CODE REVIEW

**Comment by rouault** [comment]
I think the approach of this PR, as it is done currently, should be subject for larger discussion, at least on gdal-dev and perhaps as a RFC to define a precise semantics and perhaps a dedicated API

From what I see, you use the ENCODING metadata in 2 different cases :
- in shapefile, that implement on-the-fly recoding from/to UTF-8, you always report UTF-8
- on mapinfo, which doesn't implement on-the-fly recoding, you report the native encoding of the .dat

But instead of that approach, what could potentially be done, and would be better in term of compatibility with users of the OGR API (since UTF-8 is supposed to be the pivot encoding at the API level) would be to add on-the-fly recoding from/to UTF-8 in the case of MapInfo, similarly to what is done in the shapefile case, and declare the OLCStringsAsUTF8 layer capability

**Comment by drons** [comment]
OK.
It will also be more convenient for me if all OGR API will be return data encoded in UTF-8.

I will redo my pull request and I will recode the data in the MapInfo driver in the UTF-8 encoding.

Maybe I'll add the Mitab layer creation option 'ENCODING'?
Are there any objections?

**Comment by rouault** [comment]
> I will redo my pull request and I will recode the data in the MapInfo driver in the UTF-8 encoding.

That would be great!

> Maybe I'll add the Mitab layer creation option 'ENCODING'?

Yes, makes sense too

**Comment by drons** [comment]
One more thing.
When we read shape files, the name of the encoding is converted to the standard name (OGRShapeLayer::ConvertCodePage), but when creating the layer, the internal DBF name of the encoding is used.

I suppose this should also be remade?

**Comment by drons** [comment]
Much more difficult work than I expected.
We have some limitation for now: function TABCleanFieldName can't work with UTF-8 strings.
Thus, we can't properly work with UTF-8 encoded field names.

**Comment by rouault** [comment]
I guess TABCleanFieldName() could remain restricted to ASCII characters + the current restrictions for now (I guess it is used only in the write code path ?). In a later step, could be improved to support other extended characters if needed and supported by MapInfo. The _WIN32 specific case could be completely removed I guess.

**Comment by drons** [comment]
Function MIFFile::AddFieldNative is used at read path too.
MapInfo can read/write columns names with various encoding.

I will create some additional tests and choose more complicated place for columns names recode.

**Comment by drons** [comment]
When MIF header TABCleanFieldName is used, for TAB header is not used.
see TABFile::ParseTABFileFields() and MIFFile::AddFields()
Mitab is not well known codebase for me, but i'm see significantly different ways for one job.

**Comment by drons** [comment]
Why we need to TABCleanFieldName on MIF read path (inside MIFFile::AddFieldNative)?

**Comment by rouault** [comment]
> Why we need to TABCleanFieldName on MIF read path (inside MIFFile::AddFieldNative)?

I doubt this is really intended. More that it was easier to re-use AddFieldNative() I guess

File: gdal/ogr/ogrsf_frmts/mitab/mitab_utils.cpp [hunk 1]
```diff
@@ -472,7 +472,7 @@ char *TABEscapeString(char *pszString)
 char *TABCleanFieldName(const char *pszSrcName)
 {
     char *pszNewName = CPLStrdup(pszSrcName);
-
+    //return pszNewName;
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/mitab/mitab_utils.cpp]: This commented line should be removed

File: gdal/ogr/ogrsf_frmts/mitab/mitab_miffile.cpp [hunk 2]
```diff
@@ -590,12 +601,25 @@ int  MIFFile::AddFields(const char *pszLine)
...
     int numTok = CSLCount(papszToken);
 
+    CPLString   osFieldName;
+    if( numTok > 0 )
+    {
+        osFieldName = papszToken[0];
+        if( strlen( GetEncoding() ) > 0 )
+        {
+            osFieldName.Recode( GetEncoding(), CPL_ENC_UTF8 );
+//            printf( "MIF AddFields recode from %s to %s (%s->%s)\n",
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/mitab/mitab_miffile.cpp]: This commented line should be removed

File: gdal/ogr/ogrsf_frmts/mitab/mitab_ogr_driver.cpp [hunk 3]
```diff
@@ -208,6 +208,7 @@ void RegisterOGRTAB()
     poDriver->SetMetadataItem(GDAL_DS_LAYER_CREATIONOPTIONLIST,
 "<LayerCreationOptionList>"
 "  <Option name='BOUNDS' type='string' description='Custom bounds. Expect format is xmin,ymin,xmax,ymax'/>"
+"  <Option name='ENCODING' type='string' description='to override the encoding interpretation of the DAT/MID with any encoding supported by CPLRecode or to \"\" to avoid any recoding (Neutral charset)'/>"
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/mitab/mitab_ogr_driver.cpp]: Should be documented in drv_mitab.html (with mention it is available since GDAL 2.3)

**Comment by drons** [comment]
Look like that's it!

**Comment by rouault** [comment]
Great job. Committed into trunk r39642 with minor adjustments

**Comment by drons** [comment]
Thanks!

**Pull request closed without merging**